cmd/juju/application/updateseries_test.go

```diff
@@ -0,0 +1,122 @@
+
+	"github.com/juju/juju/api/charms"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/jujuclient"
+	"github.com/juju/testing"
+)
+
+type updateSeriesSuite struct {
+	testing.IsolationSuite
+	mockAPI         *mockUpdateSeriesAPI
+	mockAPIRoot     *mockAPIConnection
+	mockCharmClient *mockCharmClient
+	store           *jujuclient.MemStore
+}
+
+var _ = gc.Suite(&updateSeriesSuite{})
+
+func (s *updateSeriesSuite) SetUpTest(c *gc.C) {
+	s.IsolationSuite.SetUpTest(c)
+	s.mockAPI = &mockUpdateSeriesAPI{Stub: &testing.Stub{}}
+	s.mockAPIRoot = &mockAPIConnection{
+		bestFacadeVersion: 5,
+		serverVersion: &version.Number{
+			Major: 1,
+			Minor: 2,
+			Patch: 3,
+		},
+	}
+	s.mockCharmClient = &mockCharmClient{
+		charmInfo: &charms.CharmInfo{
+			Meta: &charm.Meta{
+				Series: []string{"trusty", "xenial", "quantal"},
+			},
+		},
+	}
+	// Set up the current controller, and write just enough info
+	// so we don't try to refresh
+	controllerName := "test-master"
+	s.store = jujuclient.NewMemStore()
+	s.store.CurrentControllerName = controllerName
+	s.store.Controllers[controllerName] = jujuclient.ControllerDetails{}
+	s.store.Models[controllerName] = &jujuclient.ControllerModels{
+		CurrentModel: "fred/test",
+		Models: map[string]jujuclient.ModelDetails{
+			"bob/test": {"test-uuid"},
+			"bob/prod": {"prod-uuid"},
+		},
+	}
+	s.store.Accounts[controllerName] = jujuclient.AccountDetails{
+		User: "bob",
+	}
+}
+
+func (s *updateSeriesSuite) runUpdateSeries(c *gc.C, args ...string) (*cmd.Context, error) {
+	return cmdtesting.RunCommand(c, NewUpdateSeriesCommandForTest(s.mockAPI, s.mockCharmClient, s.mockAPIRoot, s.store), args...)
+}
+
+func (s *updateSeriesSuite) TestGoodPath(c *gc.C) {
+	_, err := s.runUpdateSeries(c, "ghost", "xenial")
+	c.Assert(err, jc.ErrorIsNil)
+	s.mockAPI.CheckCallNames(c, "Get", "GetCharmURL", "UpdateApplicationSeries")
+}
+
+func (s *updateSeriesSuite) TestNoArguments(c *gc.C) {
+	_, err := s.runUpdateSeries(c)
+	c.Assert(err, gc.ErrorMatches, "No application/machine and/or series specified")
+}
+
+func (s *updateSeriesSuite) TestTooManyArguments(c *gc.C) {
+	_, err := s.runUpdateSeries(c, "ghost", "xenial", "something else")
+	c.Assert(err, gc.ErrorMatches, `unrecognized args: \["something else"\]`, gc.Commentf("details: %s", errors.Details(err)))
+}
+
+func (s *updateSeriesSuite) TestSameSeries(c *gc.C) {
+	_, err := s.runUpdateSeries(c, "ghost", "trusty")
+	c.Assert(err, gc.ErrorMatches, "series trusty is already in use by ghost")
+}
+
+func (s *updateSeriesSuite) TestUnsupportedSeriesForce(c *gc.C) {
+	_, err := s.runUpdateSeries(c, "ghost", "76")
+	c.Assert(err, gc.ErrorMatches, "series 76 is not supported by the charm version in use. Use --force to override.")
+	s.mockAPI.CheckCallNames(c, "Get", "GetCharmURL")
+	_, err = s.runUpdateSeries(c, "ghost", "76", "--force")
+	c.Assert(err, jc.ErrorIsNil)
+	s.mockAPI.CheckCallNames(c, "Get", "GetCharmURL", "Get", "GetCharmURL", "UpdateApplicationSeries")
```

Comment:
We should be checking more than just call names, need to check call args as well or else how to we know the CLI is passing the expected stuff to the server side?

Instead of checking call names at all, we could just have the api mocks record what's passed in and check those struct attributes. This is what is done in other places.